### PR TITLE
Adapt the Client with the new Gridpool Trades definition and functions

### PR DESCRIPTION
**Pull request by ghost**
Following changes in the protofub definition, the `_client.py` and `_types.py` have to be adapted. 
The changes include:
* A new `Trade` class
* A new `TradeStatus` class
* A new `GridpoolTradeFilter` class
* Modifications in `OrderStatus`
* New functions to list and stream trades

### CODE REVIEW

File: py/frequenz/client/electricity_trading/_client.py [hunk 1]
```diff
@@ -91,32 +101,78 @@ async def stream_gridpool_orders(  # pylint: disable=too-many-arguments
         Returns:
             Async generator of orders.
         """
-        gridpool_filter = GridpoolOrderFilter(
-            states=order_states,
+        gridpool_order__filter = GridpoolOrderFilter(
```
Comment (matthias-wende-frequenz) [inline py/frequenz/client/electricity_trading/_client.py]: looks like there are 2 underscores after order

**Review by matthias-wende-frequenz — COMMENTED** [review]
one small comment

File: py/frequenz/client/electricity_trading/_client.py (hunk 1 above)
Comment (ghost) [reply py/frequenz/client/electricity_trading/_client.py]: uups good catch, fixed now

**Review by matthias-wende-frequenz — DISMISSED** [review]

File: RELEASE_NOTES.md [hunk 2]
```diff
@@ -12,6 +12,7 @@
 
 * Make a distinction between Order and Trade in the protobuf definitions
 * Introduction of new endpoints to retrieve gridpool trades
+* Addition of these new classes and methods to the Python client
```
Comment (daniel-zullo-frequenz) [inline RELEASE_NOTES.md]: This note is a bit vague in the sense it doesn't tell the user what classes or methods have been added. The idea of the release-notes is to briefly describe and communicate, in this case, the new feature added without the need for the user to jump into the code to understand what the feature is about

File: py/frequenz/client/electricity_trading/_client.py [hunk 3]
```diff
@@ -91,32 +101,78 @@ async def stream_gridpool_orders(  # pylint: disable=too-many-arguments
...
+        """
+        Stream gridpool trades.
+
+        Args:
+            gridpool_id: ID of the gridpool to stream trades for.
+            trade_states: List of trade states to filter for.
+            trade_id_lists: List of trade IDs to filter for.
+            market_side: Market side to filter for.
+            delivery_period: Delivery period to filter for.
+            delivery_area: Delivery area to filter for.
```
Comment (daniel-zullo-frequenz) [inline py/frequenz/client/electricity_trading/_client.py]:
```suggestion
            gridpool_id: The ID of the gridpool to stream trades for.
            trade_states: List of trade states to filter for.
            trade_id_lists: List of trade IDs to filter for.
            market_side: The market side to filter for.
            delivery_period: The delivery period to filter for.
            delivery_area: The delivery area to filter for.
```

File: py/frequenz/client/electricity_trading/_client.py [hunk 4]
```diff
@@ -91,32 +101,78 @@ async def stream_gridpool_orders(  # pylint: disable=too-many-arguments
...
                 ),
                 lambda response: OrderDetail.from_pb(response.order_detail),
             )
         return self._gridpool_orders_streams[stream_key].new_receiver()
 
+    async def stream_gridpool_trades(  # pylint: disable=too-many-arguments
+        self,
+        gridpool_id: int,
+        trade_states: list[TradeState] | None = None,
+        trade_id_lists: list[int] | None = None,
```
Comment (daniel-zullo-frequenz) [inline py/frequenz/client/electricity_trading/_client.py]: I found the parameter name confusing as it makes me think in many lists. What about `trade_ids` or something around that?

File: py/frequenz/client/electricity_trading/_client.py [hunk 5]
```diff
@@ -91,32 +101,78 @@ async def stream_gridpool_orders(  # pylint: disable=too-many-arguments
...
+        Args:
+            gridpool_id: ID of the gridpool to stream trades for.
+            trade_states: List of trade states to filter for.
+            trade_id_lists: List of trade IDs to filter for.
+            market_side: Market side to filter for.
+            delivery_period: Delivery period to filter for.
+            delivery_area: Delivery area to filter for.
+
+        Returns:
+            Async generator of trades.
```
Comment (daniel-zullo-frequenz) [inline py/frequenz/client/electricity_trading/_client.py]: I'm not sure if `generator` is the right name. I'd suggest something around `the stream of gridpool trades` or `The gridpool trades stream(er)`.

File: py/frequenz/client/electricity_trading/_types.py [hunk 6]
```diff
@@ -605,6 +601,67 @@ def to_pb(self) -> electricity_trading_pb2.OrderState.ValueType:
...
+        electricity_trading_pb2.TradeState.TRADE_STATE_APPROVAL_REQUESTED
+    )
+    """An approval has been requested."""
+
+    @classmethod
+    def from_pb(cls, trade_state: electricity_trading_pb2.TradeState.ValueType) -> Self:
+        """Convert a protobuf TradeState value to TradeState enum.
+
+        Args:
+            trade_state: Trade state to convert.
```
Comment (daniel-zullo-frequenz) [inline py/frequenz/client/electricity_trading/_types.py]:
```suggestion
            trade_state: The trade state to convert.
```

File: py/frequenz/client/electricity_trading/_client.py [hunk 7]
```diff
@@ -431,6 +487,59 @@ async def list_gridpool_orders(  # pylint: disable=too-many-arguments
...
+            trade_states: List of trade states to filter by.
+            trade_id_lists: List of trade IDs to filter by.
+            market_side: The side of the market to filter by.
+            delivery_period: The delivery period to filter by.
+            delivery_area: The delivery area to filter by.
+            max_nr_orders: The maximum number of orders to return.
+            page_token: The page token to use for pagination.
+
+        Returns:
+            The list of trades for that gridpool.
```
Comment (daniel-zullo-frequenz) [inline py/frequenz/client/electricity_trading/_client.py]:
```suggestion
            The list of trades for the given gridpool.
```

**Review by daniel-zullo-frequenz — COMMENTED** [review]
I have a few minor comments to check for, LGTM otherwise

File: py/frequenz/client/electricity_trading/_client.py (hunk 4 above)
Comment (ghost) [reply py/frequenz/client/electricity_trading/_client.py]: yea to be fair I agree with you on this one I just kept it for consistency as this was the name given in the protobuf, but I'm for changing it here

Comment (daniel-zullo-frequenz) [reply py/frequenz/client/electricity_trading/_client.py]: oh I see, then resolving this

**Review by daniel-zullo-frequenz — APPROVED** [review]

**Pull request merged**